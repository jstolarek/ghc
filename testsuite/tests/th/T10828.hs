{-# LANGUAGE TemplateHaskell, GADTs, ExplicitForAll, KindSignatures #-}

module T10828 where

import Language.Haskell.TH
import System.IO

-- data T a where
--    MkT :: a -> a -> T a
--    MkC :: forall a. (a ~ Int) => { foo :: a, bar :: b } -> T Int

$( return
   [ DataD [] (mkName "T")
           [ PlainTV (mkName "a") ]
           (Just StarT)
           [ GadtC (mkName "MkT")
                   [ (NotStrict, VarT (mkName "a"))
                   , (NotStrict, VarT (mkName "a"))]
                   ( mkName "T" )
                   [ VarT (mkName "a") ]
           , ForallC [PlainTV (mkName "a")]
                     [AppT (AppT EqualityT (VarT $ mkName "a"  ) )
                                           (ConT $ mkName "Int") ] $
             RecGadtC (mkName "MkC")
                  [ (mkName "foo", NotStrict, VarT (mkName "a"))
                  , (mkName "bar", NotStrict, VarT (mkName "b"))]
                  ( mkName "T" )
                  [ ConT (mkName "Int") ] ]
           [] ])

$( do { -- test reification
        TyConI dec <- runQ $ reify (mkName "T")
      ; runIO $ putStrLn (pprint dec) >> hFlush stdout

        -- test quoting
      ; d <- runQ $ [d|
             data T' a :: * where
                MkT' :: a -> a -> T' a
                MkC' :: forall a b. (a ~ Int) => { foo :: a, bar :: b }
                                              -> T' Int |]
      ; runIO $ putStrLn (pprint d) >> hFlush stdout
      ; return [] } )
