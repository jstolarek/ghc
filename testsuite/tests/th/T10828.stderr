data family D_0 a_1 :: * -> *
data instance D_0 GHC.Types.Int GHC.Types.Bool :: * where
    DInt_2 :: D_0 GHC.Types.Int GHC.Types.Bool
data T10828.T (a_0 :: *) where
    T10828.MkT :: a_1 -> a_1 -> T10828.T a_1
    T10828.MkC :: forall (a_2 :: *) (b_3 :: *) . Data.Type.Equality.~ a_2
                                                                      GHC.Types.Int => {T10828.foo :: a_2,
                                                                                        T10828.bar :: b_3} -> T10828.T GHC.Types.Int

data T'_0 a_1 :: * where
    MkT'_2 :: forall a_3 . a_3 -> a_3 -> T'_0 a_3
    MkC'_4 :: forall a_5 b_6 . a_5 ~ GHC.Types.Int => {foo_7 :: a_5,
                                                       bar_8 :: b_6} -> T'_0 GHC.Types.Int


TYPE SIGNATURES
TYPE CONSTRUCTORS
  type role T nominal
  data T a where
    MkT :: a -> a -> T a
    MkC :: a1 ~ Int => a1 -> b -> T Int
COERCION AXIOMS
Dependent modules: []
Dependent packages: [array-<VERSION>, base-<VERSION>, deepseq-<VERSION>,
                     ghc-boot-0.0.0.0, ghc-prim-<VERSION>, integer-<IMPL>-<VERSION>,
                     pretty-<VERSION>, template-haskell-<VERSION>]

==================== Typechecker ====================
T10828.$tcT = TyCon 0## 0## T10828.$trModule (TrNameS "T"#)
AbsBinds [] []
  {Exports: [bar <= bar
               <>]
   Exported types: bar :: ()
                   [LclIdX[[RecSel(naughty)]], Str=DmdType]
   Binds: bar = ()}
AbsBinds [] []
  {Exports: [foo <= foo
               <>]
   Exported types: foo :: ()
                   [LclIdX[[RecSel(naughty)]], Str=DmdType]
   Binds: foo = ()}
T10828.$trModule = Module (TrNameS "main"#) (TrNameS "T10828"#)
